src/primitives/hrp.rs

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+use core::cmp::Ordering;
+use core::convert::TryInto;
+use core::fmt::{self, Write};
+use core::iter::{self, FusedIterator};
+use core::slice;
+
+use crate::Case;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug)]
+pub struct Hrp {
+    /// ASCII byte values, guaranteed not to be mixed-case.
+    buf: [u8; MAX_HRP_LEN],
+    /// Number of characters currently stored in this HRP.
+    size: usize,
+}
+
+impl Hrp {
+    /// Parses the human-readable part checking it is valid as defined by [BIP-173].
+    ///
+    /// This does _not_ check that the `hrp` is an in-use HRP within Bitcoin (eg, "bc"), rather it
+    /// checks that the HRP string is valid as per the specification in [BIP-173]:
+    ///
+    /// > The human-readable part, which is intended to convey the type of data, or anything else that
+    /// > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+    /// > character having a value in the range [33-126]. HRP validity may be further restricted by
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Ok(Self::parse_and_case(hrp)?.0) }
```

Comment:
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

Would be nice to add a `parse_unchecked` function that was `const`. It could call `str::as_bytes` and just copy data over. We don't even need to mark it `unsafe` because no single-byte value is an invalid character (or at least, `char::from<u8>()` somehow deals with the ones that are).

Would save people (and us) a lot of `parse("obviouslycorrect").unwrap()` code.

Reply:
Or if you wanted to preserve invariants we could just check everything is in the 33..126 range and replace any invalid bytes (or the whole string) with `x` or something.

Reply:
I just paniced if the input is invalid.

Reply:
Interesting, I didn't know you could panic in constfns. I guess you just can't `unwrap`?

Reply:
I dunno if `unchecked` is the right word now for this :P given that the input is, in fact, checked. But I think it's fine. It signals to the user that they've gotta do their own validation.

Reply:
So, `panic` is not actually const in 1.48. I found a bug in my local CI setup that was causing the set of compilers to be sometimes collapsed into one. But I don't know why the github CI didn't catch this, and isn't catching this.

`cargo +1.48.0 test` fails right now on master.